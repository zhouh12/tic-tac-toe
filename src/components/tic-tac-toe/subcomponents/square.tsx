import React from 'react';
import { observer } from 'mobx-react';
import { SquareValue } from '../../../types/square-value-config';

export interface SquareProps {
  value: SquareValue;
  onClick: () => void;
}

// TODO: in real project, we can create a help function such as bemClassBuilder
// in order to name block or element, right now just hard-coded
const b = 'square';

/**
 * A component fills 'X' or 'O' when clicking it.
 */
export const Square: React.FC<SquareProps> = observer((props) => {
  return (
    <button className={b} onClick={props.onClick}>
      {props.value}
    </button>
  );
});
