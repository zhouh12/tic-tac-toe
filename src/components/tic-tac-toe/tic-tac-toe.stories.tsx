import React from 'react';
import { TicTacToeComponent } from './tic-tac-toe';

export default {
  title: 'Components/TicTacToeComponent',
  component: TicTacToeComponent,
};

export const Example: React.FC = () => {
  return <TicTacToeComponent buttonText="Restart Game" />;
};
