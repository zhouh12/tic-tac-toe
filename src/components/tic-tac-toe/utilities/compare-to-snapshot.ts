import { ReactElement } from 'react';
import { create, TestRendererOptions } from 'react-test-renderer';

/**
 * A helper function creates snapshot test
 */
export function compareToSnapshot(
  nextElement: ReactElement<any>,
  options?: TestRendererOptions
) {
  const tree = create(nextElement, options);
  expect(tree).toMatchSnapshot();
  tree.unmount();
}
