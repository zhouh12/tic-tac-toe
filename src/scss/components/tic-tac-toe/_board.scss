
.board {
    &__row{ 
        display: flex;
        flex-wrap: wrap;
    }
}