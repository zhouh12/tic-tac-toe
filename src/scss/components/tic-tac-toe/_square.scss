$square-background-color: #a0d2eb;
$square-border-color:#e5eaf5;
$square-border-radius: 6px;
$square-font-size:54px;
$square-font-color: #e5eaf5;
$square-font-weight: bold;
$square-line-height:1.5;
$square-height:90px;
$square-width:90px;
$square-padding:0;

.square {
    background-color: $square-background-color;
    border: 1px solid $square-border-color;
    font-size: $square-font-size;
    font-weight: $square-font-weight;
    color: $square-font-color;
    line-height: $square-line-height;
    height: $square-height;
    width: $square-width;
    padding: $square-padding;
    border-radius: $square-border-radius;
    &:focus {
        outline: none;
    }
}