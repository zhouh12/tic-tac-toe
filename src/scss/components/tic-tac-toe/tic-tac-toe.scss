@import 'board';
@import 'square';

$component-margin-top: 20px;
$component-margin-left: 20px;
$button-background-color: #13B5EA;
$button-font-size:24px;


.tic-tac-toe {
    display: flex;
    flex-direction: row;
    &__board {
        margin-left: $component-margin-left;
        margin-top: $component-margin-top;
    }
    &__info {
        margin-left: $component-margin-left;
        margin-top: $component-margin-top;
    }
    &__button {
        background-color: $button-background-color;
        color: #FFF;
        font-size: $button-font-size;
    }
}