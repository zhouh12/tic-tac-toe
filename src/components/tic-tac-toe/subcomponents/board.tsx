import React, { ReactNode } from 'react';
import { observer } from 'mobx-react';
import { SquareValue } from '../../../types/square-value-config';
import { Square } from './square';

export interface BoardProps {
  squares: SquareValue[];
  onClick: (index: number) => void;
}

// TODO: in real project, we can create a help function such as bemClassBuilder
// in order to name block or element, right now just hard-coded
const b = 'board__row';

/**
 * A component renders 9 squares.
 */
export const Board: React.FC<BoardProps> = observer((props) => {
  const renderSquare = (i: number): ReactNode => {
    return <Square value={props.squares[i]} onClick={() => props.onClick(i)} />;
  };

  return (
    <div>
      <div className={b}>
        {renderSquare(0)}
        {renderSquare(1)}
        {renderSquare(2)}
      </div>
      <div className={b}>
        {renderSquare(3)}
        {renderSquare(4)}
        {renderSquare(5)}
      </div>
      <div className={b}>
        {renderSquare(6)}
        {renderSquare(7)}
        {renderSquare(8)}
      </div>
    </div>
  );
});
