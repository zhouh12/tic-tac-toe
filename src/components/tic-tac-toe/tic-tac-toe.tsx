import React, { useState } from 'react';
import { SquareValue } from '../../types/square-value-config';
import { calculateWinner } from './utilities/calculate-winner';
import { Board } from './subcomponents/board';
import { TicTacToeProps } from './types/props';
import { observer } from 'mobx-react';

// TODO: in real project, we can create a help function such as bemClassBuilder
// in order to name block or element, right now just hard-coded
const b = 'tic-tac-toe';
const bBoard = 'tic-tac-toe__board';
const bInfo = 'tic-tac-toe__info';
const bButton = 'tic-tac-toe__button';
/**
 * A top-level component displays the game.
 */
export const TicTacToeComponent: React.FC<TicTacToeProps> = observer(
  (props) => {
    const [isNext, setIsNext] = useState<boolean>(true);
    const [stepNumber, setStepNumber] = useState<number>(0);
    const [history, setHistory] = useState<{ squares: SquareValue[] }[]>([
      {
        squares: Array(9).fill(null),
      },
    ]);

    const handleClick = (index: number) => {
      const newHistory = history.slice(0, stepNumber + 1);
      const current = newHistory[newHistory.length - 1];
      const squares = current.squares.slice();
      if (calculateWinner(squares) || squares[index]) {
        return;
      }
      squares[index] = isNext ? 'X' : 'O';
      setHistory(
        newHistory.concat([
          {
            squares: squares,
          },
        ])
      );
      setStepNumber(newHistory.length);
      setIsNext(!isNext);
    };

    const current = history[stepNumber];
    const winner = calculateWinner(current.squares);

    let status;
    if (winner) {
      status = 'Winner: ' + winner;
    } else {
      status = 'Next player: ' + (isNext ? 'X' : 'O');
    }

    return (
      <div className={b}>
        <div className={bBoard}>
          <div>
            <Board squares={current.squares} onClick={(i) => handleClick(i)} />
          </div>
        </div>
        <div className={bInfo}>
          <div>{status}</div>
          <div>
            <button
              className={bButton}
              onClick={() => {
                setStepNumber(0);
              }}
            >
              {props.buttonText}
            </button>
          </div>
        </div>
      </div>
    );
  }
);
