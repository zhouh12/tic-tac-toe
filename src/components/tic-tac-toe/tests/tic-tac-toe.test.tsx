import { compareToSnapshot } from '../utilities/compare-to-snapshot';
import React from 'react';
import { TicTacToeComponent } from '../tic-tac-toe';
import { Board } from '../subcomponents/board';
import { shallow } from 'enzyme';

describe('TicTacToeComponent', () => {
  const buttonText = 'Restart Game';

  it('renders correctly', () => {
    compareToSnapshot(<TicTacToeComponent buttonText={buttonText} />);
  });

  it('should call handleClick when square is clicked', () => {
    const onClickFunc = jest.fn();

    const wrapper = shallow(<TicTacToeComponent buttonText={buttonText} />);
    wrapper.find(Board).simulate('click');
    expect(onClickFunc).toHaveBeenCalledTimes(0);
    wrapper.unmount();
  });
});
