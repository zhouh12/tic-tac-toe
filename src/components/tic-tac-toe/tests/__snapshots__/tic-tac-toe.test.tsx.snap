// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TicTacToeComponent renders correctly 1`] = `
<div
  className="tic-tac-toe"
>
  <div
    className="tic-tac-toe__board"
  >
    <div>
      <div>
        <div
          className="board__row"
        >
          <button
            className="square"
            onClick={[Function]}
          />
          <button
            className="square"
            onClick={[Function]}
          />
          <button
            className="square"
            onClick={[Function]}
          />
        </div>
        <div
          className="board__row"
        >
          <button
            className="square"
            onClick={[Function]}
          />
          <button
            className="square"
            onClick={[Function]}
          />
          <button
            className="square"
            onClick={[Function]}
          />
        </div>
        <div
          className="board__row"
        >
          <button
            className="square"
            onClick={[Function]}
          />
          <button
            className="square"
            onClick={[Function]}
          />
          <button
            className="square"
            onClick={[Function]}
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="tic-tac-toe__info"
  >
    <div>
      Next player: X
    </div>
    <div>
      <button
        className="tic-tac-toe__button"
        onClick={[Function]}
      >
        Restart Game
      </button>
    </div>
  </div>
</div>
`;
